File: tests/integration/keyof-operator-test/keyof-operator-test.cpp
#include "keyof-operator-test.h"

using namespace js;

js::any getProperty(Person person, PersonKeys key) {
    return person[key];
}

template<typename T>
js::array<js::string> getKeys(T obj) {
    return js::object.keys(obj);
}

auto main() {
    const Person person = []() {
          js::object obj_temp_0;
          obj_temp_0.set("name", "John"_S);
          obj_temp_0.set("age", js::number(30));
          obj_temp_0.set("email", "john@example.com"_S);
          return js::any(obj_temp_0);
        }();
    const js::any name = getProperty(person, "name"_S);
    const js::any age = getProperty(person, "age"_S);
    const js::any keys = getKeys(person);
    js::console.log("Person keys:"_S, keys);
    const Point point = []() {
          js::object obj_temp_1;
          obj_temp_1.set("x", js::number(10));
          obj_temp_1.set("y", js::number(20));
          return js::any(obj_temp_1);
        }();
    const js::any pointKeys = getKeys(point);
    js::console.log("Point keys:"_S, pointKeys);
}

File: tests/integration/keyof-operator-test.ts
interface Person {
  name: string;
  age: number;
  email: string;
}

// keyof Person should resolve to "name" | "age" | "email"
type PersonKeys = keyof Person;

// Function that accepts only valid keys of Person
function getProperty(person: Person, key: PersonKeys): any {
  return person[key];
}

// Using keyof with generic constraints
function getKeys<T>(obj: T): (keyof T)[] {
  return Object.keys(obj) as (keyof T)[];
}

// keyof with type alias
type Point = { x: number; y: number };
type PointKeys = keyof Point; // "x" | "y"

// keyof with index signatures
interface StringMap {
  [key: string]: string;
}
type StringMapKeys = keyof StringMap; // string

// keyof with number index
interface NumberArray {
  [index: number]: string;
}
type NumberArrayKeys = keyof NumberArray; // number

// Test usage
function main() {
  const person: Person = {
    name: "John",
    age: 30,
    email: "john@example.com",
  };

  // Valid key access
  const _name = getProperty(person, "name");
  const _age = getProperty(person, "age");

  // Get all keys
  const keys = getKeys(person);
  console.log("Person keys:", keys);

  // Iterate over keys
  const point: Point = { x: 10, y: 20 };
  const pointKeys = getKeys(point);
  console.log("Point keys:", pointKeys);
}

main();
